[DAPS-1515] Integrate repository type system into core endpoints

AronPerez opened the pull request:
## Summary
- Integrates the new repository type system into existing API endpoints
- Adds type-aware operations for data uploads and allocations
- Production code changes that enable repository type features

## Changes

### data_router.js
- Add check for repository data operation support before uploads
- Prevent data uploads to metadata-only repositories
- Return appropriate error messages with repository type info

### repo_router.js  
- Update allocation create/delete endpoints to use RepositoryOps
- Add repository type awareness to allocation operations
- Support different response formats based on repository type

### support.js
- Add new error codes for repository type operations
- Add helper functions for repository type validation

### tasks.js
- Update task creation to handle different repository types
- Add type-specific task handling logic

## Test plan
- [x] Endpoints maintain backward compatibility
- [ ] Data uploads blocked for metadata-only repos
- [ ] Allocations work correctly for all repo types
- [ ] Integration tests pass

## Context
This is PR 3 of 4 that breaks down PR #1538.
- Previous: Legacy Repo updates (#1542)
- Next: Repository tests

JoshuaSBrown commented on core/database/foxx/api/data_router.js:
```diff
@@ -90,6 +92,24 @@ function recordCreate(client, record, result) {
+                throw [
+                    g_lib.ERR_INVALID_OPERATION,
+                    "Data uploads not supported for metadata-only repository",
+                    {
+                        repo_type: repository.type,
+                        repo_id: repository.data._id,
+                    },
```
I like that this information is being included in the exception.

JoshuaSBrown commented on core/database/foxx/api/repo_router.js:
```diff
@@ -739,13 +759,29 @@ router
-                        subject_id,
-                    );
+                    // Find the repository using the new type system
+                    var findResult = RepositoryOps.find(req.queryParams.repo);
+                    if (!findResult.ok) {
+                        throw [findResult.error.code, findResult.error.message];
+                    }
+
+                    var repository = findResult.value;
+
+                    // Check permissions
+                    var permResult = RepositoryOps.checkPermission(repository, client._id, "admin");
+                    if (!permResult.ok || !permResult.value) {
+                        throw g_lib.ERR_PERM_DENIED;
+                    }
+
+                    // Delete allocation using the new system
+                    var deleteResult = RepositoryOps.deleteAllocation(repository, subject_id);
+
+                    if (!deleteResult.ok) {
+                        throw [deleteResult.error.code, deleteResult.error.message];
+                    }
 
-                    res.send(result);
+                    // Return the new response format
+                    res.send(deleteResult.value);
```
I love how readable this is.

JoshuaSBrown commented on core/database/foxx/api/repo_router.js:
```diff
@@ -739,13 +759,29 @@ router
+                    }
+
+                    // Delete allocation using the new system
+                    var deleteResult = RepositoryOps.deleteAllocation(repository, subject_id);
```
DeleteAllocation does not check anywhere that the subject actually exists.

The below is from the initial taskInitAllocDelete call.
```
if (!g_db._exists(a_subject_id))                                                                                                                                       throw [g_lib.ERR_NOT_FOUND, "Subject, '" + a_subject_id + "', does not exist"];          
```

JoshuaSBrown replied:
This function is missing from the branch when I check it out so hard to follow if it is checking.

```
g_tasks.repoAllocationDeleteTask
```

but is called in foxx/api/globus.js

AronPerez commented:
@JoshuaSBrown Changes complete:

- Dynamic error messages using repository.type
- Added subject existence check at L776-779
- Extracted validation helper to reduce duplication

PTAL

AronPerez commented:
@JoshuaSBrown Addressed all feedback:

- Dynamic error messages now use `repository.type` for context
- Added subject existence validation at L776-779 
- Extracted common operations to `RepositoryOps` to reduce duplication

Changes committed and CI passing (ignoring shfmt per team decision).

Ready for re-review.

JoshuaSBrown commented on core/database/foxx/api/data_router.js:
```diff
@@ -90,6 +92,24 @@ function recordCreate(client, record, result) {
 
+        // Check if repository supports data operations
+        var findResult = RepositoryOps.find(repo_alloc._to);
+        if (findResult.ok) {
```
What happens when findResult is not ok, shouldn't we be throwing?

JoshuaSBrown commented on core/database/foxx/api/repo_router.js:
```diff
@@ -7,6 +7,21 @@ const joi = require("joi");
+// Helper function to prepare repository data for saving
+const prepareRepoData = (obj) => {
+    // Ensure paths end with / for saving
+    if (obj.path && !obj.path.endsWith("/")) {
+        obj.path += "/";
+    }
```
You are making sure that the .path ends with "/", in the original code, it is also checking to make sure that it begins with "/". Is that check being done somewhere else?

JoshuaSBrown replied:
I see it is in the validation.js file.

JoshuaSBrown commented on core/database/foxx/api/repo_router.js:
```diff
@@ -722,13 +759,34 @@ router
+
+                    // Check permissions
+                    var permResult = RepositoryOps.checkPermission(repository, client._id, "admin");
+                    if (!permResult.ok || !permResult.value) {
+                        throw g_lib.ERR_PERM_DENIED;
+                    }
+
+                    // Check if subject exists
+                    if (!g_db._exists(subject_id)) {
+                        throw [g_lib.ERR_NOT_FOUND, "Subject not found: " + subject_id];
```
It would be clearer if we include in the exeception, that the allocation delete request failed, for audit reasons, in cases around authorization it is really useful to know who attempted to do what.

JoshuaSBrown commented on core/database/foxx/api/repository/metadata.js:
```diff
@@ -0,0 +1,136 @@
+};
+
+// Delete allocation from metadata repository (direct/synchronous)
+const deleteAllocation = (repoData, subjectId) => {
+    if (!subjectId || typeof subjectId !== "string") {
+        return Result.err({
+            code: g_lib.ERR_INVALID_PARAM,
+            message: "Subject ID is required for allocation deletion",
+        });
+    }
+
+    try {
+        // For metadata-only repos, just remove the database record
+        // No actual storage deallocation needed
+        const result = {
+            repo_id: repoData._id,
+            subject: subjectId,
+            status: "completed",
+            message: "Metadata allocation removed",
+        };
+
+        return Result.ok(createAllocationResult(ExecutionMethod.DIRECT, result));
+    } catch (e) {
+        return Result.err({
+            code: g_lib.ERR_INTERNAL_FAULT,
+            message: `Failed to delete metadata allocation: ${e.message}`,
+        });
+    }
+};
```
Maybe you are implementing that in a later PR? It doesn't look like you are removing the allocation document here.

JoshuaSBrown commented on core/database/foxx/api/repository/metadata.js:
```diff
@@ -0,0 +1,136 @@
+        // For metadata-only repos, allocations are just database records
+        // No actual storage allocation happens
+        const allocation = {
+            _key: `alloc_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
```
```
_id: 'alloc/403756027'
```

AronPerez replied:
Just a placeholder, need to update that

JoshuaSBrown commented on core/database/foxx/api/repository/metadata.js:
```diff
@@ -0,0 +1,136 @@
+            subject: allocation.subject,
+            size: allocation.size,
+            path: allocation.path,
+            status: "completed",
```
This is also not consistent with what is in the tasks.js and globus files. If you look in the support.js file you will see that there are these entried:

```
    obj.TS_BLOCKED = 0;
    obj.TS_READY = 1;
    obj.TS_RUNNING = 2;
    obj.TS_SUCCEEDED = 3;
    obj.TS_FAILED = 4;
    obj.TS_COUNT = 5;
```

They are not returned as strings but as codes. Either we need to be consistent with what is in here or also change what status is being returned by the globus status entry.
